### PR TITLE
refactor: migrate OC_EventSource to dependency injection

### PR DESCRIPTION
**Pull request by kesselb**
* Resolves: # 

## Summary

Test: Increase the version number in version.php and visit Nextcloud. 
The update should work as before.

Tip: Set a breakpoint in OC_EventSource.send and watch the request in your browser.

image

## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/public/IServerContainer.php [hunk 1]
```diff
@@ -395,15 +395,6 @@ public function getSearch();
...
 	public function getCertificateManager();
 
-	/**
-	 * Create a new event source
-	 *
-	 * @return \OCP\IEventSource
-	 * @since 8.0.0
-	 * @deprecated 20.0.0 have it injected or fetch it through \Psr\Container\ContainerInterface::get
-	 */
-	public function createEventSource();
```
Comment (ChristophWurst) [inline lib/public/IServerContainer.php:405]: Fine to be removed because deprecated but needs docs

**Review by ChristophWurst — APPROVED** [review]

**Comment by kesselb** [comment]
I overlooked that createEventSource would always create a new object. 
Updated the container definition and set $shared = false for the server. 

Can't say if a new instance makes sense, but that was the old behavior.

**Comment by kesselb** [comment]
Example: https://github.com/Rello/audioplayer/pull/586

**Review by ChristophWurst — APPROVED** [review]

**Review by come-nc — COMMENTED** [review]
In this instance I find it clearer to use createEventSource, as it makes it clear this is not a singleton. What are the pros of DI for non-singletons?

**Comment by kesselb** [comment]
> In this instance I find it clearer to use createEventSource, as it makes it clear this is not a singleton. What are the pros of DI for non-singletons?

I find that a valid point. My motivation was to inject OC::$server->getRequest() via DI. 

@ChristophWurst you deprecated it a while ago. Are you okay to keep `createEventSource`?

Should we also add createEventSource to https://github.com/nextcloud/server/blob/master/lib/public/Server.php so apps can use it without using a private api?

**Comment by icewind1991** [comment]
Might be worth adding an `EventSourceFactory` that can be injected. Same pattern as we use for http clients and others

**Comment by ChristophWurst** [comment]
Yes let's go with a injectable and mockable factory instead of a static function

**Review by ChristophWurst — APPROVED** [review]
Clean

**Review by icewind1991 — APPROVED** [review]

**Pull request merged**